src/frequenz/sdk/util/ringbuffer.py

```diff
@@ -0,0 +1,412 @@
+# License: MIT
+# Copyright © 2023 Frequenz Energy-as-a-Service GmbH
+
+"""Ringbuffer implementation with focus on time & memory efficiency."""
+
+from __future__ import annotations
+
+from copy import deepcopy
+from datetime import datetime, timedelta
+from typing import Generic, Sequence, TypeVar
+
+import numpy as np
+
+T = TypeVar("T")
+
+Container = TypeVar("Container", list, np.ndarray)
+
+
+class OrderedRingBuffer(Generic[T, Container]):
+    """Time aware ringbuffer that keeps its entries sorted by time."""
+
+    def __init__(
+        self,
+        buffer: Container,
+        sampling_period_s: float,
+        time_index_alignment: datetime = datetime(1, 1, 1),
+    ) -> None:
+        """Initialize the time aware ringbuffer.
+
+        Args:
+            buffer: Instance of a buffer container to use internally.
+            sampling_period_s: Resampling period in seconds.
+            time_index_alignment: Arbitary point in time used to align
+                timestamped data with the index position in the buffer.
+                Used to make the data stored in the buffer align with the
+                beginning and end of the buffer borders.
+
+                For example, if the `time_index_alignment` is set to
+                "0001-01-01 12:00:00", and the `sampling_period_s` is set to
+                3600 (1 hour) and the length of the buffer is 24, then the data
+                stored in the buffer could correspond to the time range from
+                "2022-01-01 12:00:00" to "2022-01-02 12:00:00" (date choosen
+                arbitrarily here).
+        """
+        self._buffer = buffer
+        self._sampling_period_s = sampling_period_s
+        self._time_index_alignment = time_index_alignment
+
+        self._missing_windows = []
+        self._datetime_newest = datetime.min
+        self._datetime_oldest = datetime.max
+        self._time_range = timedelta(
+            seconds=(len(self._buffer) - 1) * self._sampling_period_s
+        )
+
+    def normalize_timestamp(self, timestamp: datetime) -> datetime:
```

Comment:
is this for only for internal use?  Maybe needs a `_` prefix?

Reply:
can be private